src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs

```diff
@@ -84,91 +205,153 @@ thunkInfoToIsNF :: ThunkInfo -> Bool
+--
+-- > deriving via UseIsNormalFormNamed "T" T instance NoUnexpecedThunks T
+newtype UseIsNormalFormNamed (name :: Symbol) a = UseIsNormalFormNamed a
 
--- | Check if argument contains no unexpected thunks by converting to generic
--- representation first
+-- | Newtype wrapper for use with @deriving via@ to check for WHNF only
+--
+-- For some types we don't want to check for nested thunks, and we only want
+-- check if the argument is in WHNF, not in NF. A typical example are functions;
+-- see the instance of @(a -> b)@ for detailed discussion. This should be used
+-- sparingly.
 --
--- NOTE: if it's in WHNF then we can safely translate to generic value without
--- forcing anything.
-genericWhnfNoUnexpectedThunks :: forall a.
-                                 (Generic a, GWhnfNoUnexpectedThunks (Rep a))
-                              => [String] -> a -> IO ThunkInfo
-genericWhnfNoUnexpectedThunks ctxt x =
-    -- Force the result of @from@ to WHNF: we are not interested in thunks
-    -- that arise from the translation to the generic representation.
-    let fp :: Rep a x
-        !fp = from x
-    in gWhnfNoUnexpectedThunks ctxt fp
+-- Example:
+--
+-- > deriving via DontCheckForThunks "T" T instance NoUnexpectedThunks T
+newtype OnlyCheckIsWHNF (name :: Symbol) a = OnlyCheckIsWHNF a
 
--- | Short-circuit a list of checks
-allNoUnexpectedThunks :: [IO ThunkInfo] -> IO ThunkInfo
-allNoUnexpectedThunks = go
-  where
-    go :: [IO ThunkInfo] -> IO ThunkInfo
-    go []     = return NoUnexpectedThunks
-    go (a:as) = do
-        nf <- a
-        case nf of
-          NoUnexpectedThunks  -> go as
-          UnexpectedThunk nfo -> return $ UnexpectedThunk nfo
+-- | Newtype wrapper for values that should be allowed to be a thunk
+--
+-- This should be used /VERY/ sparingly, and should /ONLY/ be used on values
+-- (or, even rarer, types) which you are /SURE/ cannot retain any data that they
+-- shouldn't. Bear in mind allowing a value of type @T@ to be a thunk might
+-- cause a value of type @S@ to be retained if @T@ was computed from @S@.
+newtype AllowThunk a = AllowThunk a
+
+{-------------------------------------------------------------------------------
+  Internal: instances for the deriving-via wrappers
+-------------------------------------------------------------------------------}
+
+instance Typeable a => NoUnexpectedThunks (UseIsNormalForm a) where
+  showTypeOf _ = show $ typeRep (Proxy @a)
+  whnfNoUnexpectedThunks = noUnexpectedThunksUsingNormalForm
+
+instance KnownSymbol name => NoUnexpectedThunks (UseIsNormalFormNamed name a) where
+  showTypeOf _ = symbolVal (Proxy @name)
+  whnfNoUnexpectedThunks = noUnexpectedThunksUsingNormalForm
+
+instance KnownSymbol name => NoUnexpectedThunks (OnlyCheckIsWHNF name a) where
+  showTypeOf _ = symbolVal (Proxy @name)
+  whnfNoUnexpectedThunks _ _ = return NoUnexpectedThunks
+
+instance NoUnexpectedThunks (AllowThunk a) where
+  showTypeOf _ = "<never used since never fails>"
```

Comment:
`error`?

Reply:
If we _do_ see this in a stack, having the actual string here (plus the full context) will give us better debugging info than using `error`.